### ISSUE TITLE
Derivatives of the curve function

### ISSUE THREAD

**Issue by e-** [lmcinnes/umap#72]
Hello. Thank you for the great work. I was a little bit unsure about the derivatives of a curve in the `optimize_layout` function. From what I saw, UMAP uses an exponentially decaying curve which is `1 / (1 + a  * x ** (2 * b))`. The derivatives of the curve are used in optimization as follows:

```py
grad_coeff = (-2.0 * a * b * pow(dist_squared, b - 1.0))
grad_coeff /= (a * pow(dist_squared, b) + 1.0)
```

However, from my calculation, derivatives should be:

```py
grad_coeff = (-2.0 * a * b * pow(dist_squared, b - 0.5))
# -0.5 instead of -1 since the distance is already squared
grad_coeff /= (a * pow(dist_squared, b) + 1.0) ** 2
# must be squared
```

Please correct me if I was wrong.

My second question is about negative sampling. The `optimize_layout` function uses the following gradient coefficient:

```py
grad_coeff = (2.0 * gamma * b)
grad_coeff /= (0.001 + dist_squared) * (a * pow(dist_squared, b) + 1)
```

I would like to know how the coefficient was derived. Thanks.

**Comment by lmcinnes** [comment]
The gradients were computed fairly directly for log(membership_strength)
and log(1-membership_strength). You can simply plug the relevant equations
into wolfram alpha:

http://www.wolframalpha.com/input/?i=grad+log(1%2F(1+%2B+a+*+sqrt(x%5E2+%2B+y%5E2)%5E(2*b))
http://www.wolframalpha.com/input/?i=grad+log(1+-+1%2F(1+%2B+a+*+sqrt(x%5E2+%2B+y%5E2)%5E(2*b))

The negative gradient has a corrected (0.001) to cope with NaN's that occur
for zero distances. Other than that it is pretty straightforward I believe.
Have I missed something obvious?

**Comment by lmcinnes** [comment]
The objective is the combination of cross entropy with singular set functor. If you expand out the cross entropy and ignore terms entirely in terms of the high dimensional membership strengths then you get down to just positive and negative sample terms that are logs of the relevant functions.
